### ISSUE TITLE
Improve/Rework msgpackR extension integration

### ISSUE THREAD

**Issue by tortmayr** [eclipse-theia/theia#11941]
### Feature Description:
The current handling of msgpackR extension in the RPC infrastructure is sub optiomal an needs some polishing.
For details see https://github.com/eclipse-theia/theia/pull/11261#issuecomment-1332454885

**Comment by paul-marechal** [comment]
The current implementation seems to install global extensions for each class we may try to serialize. But being global it means we need to find non-conflicting ids for each extensions, and honestly I don't think it would scale really well for arbitrary types. This mechanism seems useful for "native" types, e.g. `BigInt`, maybe `Map`.

Now, binary serializer or not, I was under he impression that we should always make sure to "dry up" our objects before passing them to the serializer: Only handle JSON-like structures between our remote components, which in turn may or may not revive the data before propagating it further.

**Comment by tortmayr** [comment]
>The current implementation seems to install global extensions for each class we may try to serialize. But being global it means we need to find non-conflicting ids for each extensions, and honestly I don't think it would scale really well for arbitrary types. This mechanism seems useful for "native" types, e.g. BigInt, maybe Map.

I agree and share your concerns regarding the msgpackR extensions. In its current global form they definitely don`t scale well for arbitrary types. 

>Now, binary serializer or not, I was under he impression that we should always make sure to "dry up" our objects before passing them to the serializer: Only handle JSON-like structures between our remote components, which in turn may or may not revive the data before propagating it further.

Looking at the  implementations of the old Theia extension RPC (json-rpc) protocol and the old plugin RPC protocol this assumption does not quite hold. It definitely was the case in the extension RPC protocol. Here, everything was tranfsered as dehydrated json.

The plugin RPC protocol,however uses dedicated[ reviver & replacer functions ](https://github.com/eclipse-theia/theia/blob/cb0778617e8726fa51c920208b7468149753a3fb/packages/plugin-ext/src/common/rpc-protocol.ts#L390) to customize the serialization/deserialization of some base types (URI, VscodeURI, BinaryBuffer, Range directly.
In general, I wonder how postprocessing received JSON data before propagating it further impacts the performance. 
e.g. It would be necessary to do a deep, recursive check and iterate over all properties of the JSON-like structure to see whether it holds an object that requires dehydration. This sounds like a performance bottle neck to me especially when thinking about large object and/or arrarys.
